File: ps2_q1.m
% Define the symbolic variables
syms a11 a12 a21 a22 lambda

%%% PART 1
fprintf('Question 1: Solve the general matrix\n')

% Define the matrix
A = [a11, a12; a21, a22];

% Compute the characteristic polynomial
charPoly = det(A - lambda * eye(2));

% Solve for eigenvalues
eigenvalues = solve(charPoly, lambda);

% Compute eigenvectors for each eigenvalue
eigenvectors = [];
for i = 1:length(eigenvalues)
    eigenvectors = [eigenvectors, null(A - eigenvalues(i) * eye(2))];
end

% Display eigenvalues in LaTeX
disp('Eigenvalues:');
for i = 1:length(eigenvalues)
    disp(latex(eigenvalues(i)));
end

% Display eigenvectors in LaTeX
disp('Eigenvectors:');
for i = 1:length(eigenvectors)
    disp(latex(eigenvectors(:, i)));
end

%%% PART 2
fprintf('\nQuestion 2: a_21 = 0\n')

% Define the matrix
A = [a11, a12; 0, a22];

% Compute the characteristic polynomial
charPoly = det(A - lambda * eye(2));

% Solve for eigenvalues
eigenvalues = solve(charPoly, lambda);

% Compute eigenvectors for each eigenvalue
eigenvectors = [];
for i = 1:length(eigenvalues)
    eigenvectors = [eigenvectors, null(A - eigenvalues(i) * eye(2))];
end

% Display eigenvalues in LaTeX
disp('Eigenvalues:');
for i = 1:length(eigenvalues)
    disp(latex(eigenvalues(i)));
end

% Display eigenvectors in LaTeX
disp('Eigenvectors:');
for i = 1:length(eigenvectors)
    disp(latex(eigenvectors(:, i)));
end

%%% PART 3
fprintf('\nQuestion 3: a_12 = 0\n')

% Define the matrix
A = [a11, 0; a21, a22];

% Compute the characteristic polynomial
charPoly = det(A - lambda * eye(2));

% Solve for eigenvalues
eigenvalues = solve(charPoly, lambda);

% Compute eigenvectors for each eigenvalue
eigenvectors = [];
for i = 1:length(eigenvalues)
    eigenvectors = [eigenvectors, null(A - eigenvalues(i) * eye(2))];
end

% Display eigenvalues in LaTeX
disp('Eigenvalues:');
for i = 1:length(eigenvalues)
    disp(latex(eigenvalues(i)));
end

% Display eigenvectors in LaTeX
disp('Eigenvectors:');
for i = 1:length(eigenvectors)
    disp(latex(eigenvectors(:, i)));
end

File: ps2_q1.py
import sympy as sp

### PART 1: Solve the general matrix:

# Define the symbols
a11, a12, a21, a22, lam = sp.symbols('a11 a12 a21 a22 lam')
x, y = sp.symbols('x y')  # For eigenvectors

# Define the matrix A
A = sp.Matrix([[a11, a12],
            [0, a22]])

# Compute the characteristic polynomial det(A - lambda*I) = 0
I = sp.eye(2)  # Identity matrix of size 2x2
char_poly = (A - lam * I).det()

# Solve the characteristic equation for eigenvalues
eigenvalues = sp.solve(sp.Eq(char_poly, 0), lam)

# Solve for eigenvectors for each eigenvalue
eigenvectors = {}
for eigenvalue in eigenvalues:
    null_space = (A - eigenvalue * I).nullspace()
    eigenvectors[eigenvalue] = null_space

# Generate LaTeX strings for the eigenvalues
eigenvalue_latex = [sp.latex(ev) for ev in eigenvalues]

# Generate LaTeX strings for eigenvectors
eigenvector_latex = {sp.latex(ev): [sp.latex(v) for v in vectors] for ev, vectors in eigenvectors.items()}

# Display eigenvalues and eigenvectors in LaTeX
print('Question 1:', '\n',eigenvalue_latex, '\n', eigenvector_latex)

